src/builtin_string.cpp

```diff
@@ -877,12 +880,102 @@ class pcre2_matcher_t : public string_matcher_t {
+            wchar_t name[];
+#else
+            char32_t name[];
+#endif  // WCHAR_T_BITS
+#endif  // PCRE2_CODE_UNIT_WIDTH
+        };
+
+        auto &vars = this->parser.vars();
+        PCRE2_SIZE *ovector = pcre2_get_ovector_pointer(regex.match);
+        auto *names = static_cast<name_table_entry_t *>((void *)(name_table));
+        for (uint32_t i = 0; i < name_count; ++i) {
+            auto &name_entry = names[i * name_entry_size];
+
+            if (first_time) {
+                // Make sure we clear out any existing values matching the named groups, because
+                // we'll be appending instead of overwriting from here on out.
+                vars.set_empty(name_entry.name, ENV_DEFAULT);
+            }
+
+            // A named group may actually correspond to multiple group numbers, each of which
+            // might have to be enumerated.
+            PCRE2_SPTR first = nullptr;
+            PCRE2_SPTR last = nullptr;
+            int entry_size = pcre2_substring_nametable_scan(
+                regex.code, (PCRE2_SPTR)(name_entry.name), &first, &last);
+            if (entry_size <= 0) {
+                debug(1, L"PCRE2 failure retrieving named matches");
+                continue;
+            }
+
+            auto append_value = [&](wcstring &&value) {
+                auto &var = vars.get(name_entry.name, ENV_DEFAULT).value();
+                wcstring_list_t values{};
+                var.to_list(values);
+                values.emplace_back(value);
+                vars.set(name_entry.name, ENV_DEFAULT, values);
+            };
+
+            if (!match_found) {
+                if (!opts.all) {
+                    append_value(L"");
+                }
+                continue;
+            }
+
+            bool value_found = false;
+            for (auto group_ptr = first; group_ptr <= last; group_ptr += entry_size) {
+                int group_num = group_ptr[0];
+
+                PCRE2_SIZE *capture = ovector + (2 * group_num);
+                PCRE2_SIZE begin = capture[0];
+                PCRE2_SIZE end = capture[1];
+
+                if (begin != PCRE2_UNSET && end != PCRE2_UNSET && end >= begin) {
+                    append_value(haystack.substr(begin, end - begin));
+                    value_found = true;
+                    break;
+                }
+            }
+
+            // We don't have a way of having empty values in the middle of a multi-entry fish array,
+            // so we compromise by leaving the value unset in the case of !opts.all but assign an
+            // empty value otherwise. (opts.all is always true if !first_time)
+            if (!value_found && opts.all) {
+                append_value(wcstring{});
+            }
+        }
+    }
+
    public:
     pcre2_matcher_t(const wchar_t *argv0_, const wcstring &pattern, const options_t &opts,
-                    io_streams_t &streams)
+                    io_streams_t &streams, parser_t &parser_)
```

Comment:
Usually member variables have the underscore. So the parameter would be `parser` and the member variable would be `parser_`.

Reply:
@ridiculousfish I agree, but was sticking to the local convention (see parameter `argv0_`). Would you like me to change them all?